### ISSUE TITLE
Code for domain adaptation experiment?

### ISSUE THREAD

**Issue by alex-lew** [ChengtaoLi/dan#1]
Hi Chengtao,

First off, thanks for releasing this code! 

I was wondering if you had plans to release your code for the domain adaptation experiments, particularly the MNIST/MNIST-M experiment you mention in your paper. If not, do you think you could briefly describe the architecture you use for your Deep Mean Encoder and distributional adversary? (Or do you use the same architecture for that adversary as in line 271 [here](https://github.com/ChengtaoLi/dan/blob/master/mnist/model_mnist_mlp.py)?) I assume learning rate and optimizer parameters for the distributional adversary is the same as for the regular DANN adversary.

Thanks so much!

**Comment by ChengtaoLi** [comment]
Hi, thank you very much for your interest in our work!

We would love to release the code if that is helpful to you. The code for DANN experiment is largely developed upon (https://github.com/pumpikano/tf-dann) and the distributional adversary architecture is the same as the single sample adversary architecture there. Please stay tuned and we will let you know when it's there.

Thanks!
Chengtao

**Comment by alex-lew** [comment]
Thank you! One clarification. You say that your distributional adversary architecture is the same as the single sample adversary architecture in the tf-dann repo; but when do you take the mean of the batch? In your code in this repo, I noticed you usually have a fully connected layer and then a `tf.reduce_mean` across sample points (this is the "deep mean encoder" from the paper). In the DANN experiment, do you _add_ a deep mean encoding layer before the domain predictor, or do you _modify_ the first layer of the domain predictor to take the mean of its hidden units (e.g. as in code below)?

```python
with tf.variable_scope('distributional_domain_predictor'):
     # Flip the gradient when backpropagating through this operation
     feat = flip_gradient(self.feature, self.l)
            
     d_W_fc0 = weight_variable([7 * 7 * 48, 100])
     d_b_fc0 = bias_variable([100])
     # add reduce_mean here?
     d_h_fc0 = tf.reduce_mean(tf.nn.relu(tf.matmul(feat, d_W_fc0) + d_b_fc0), axis=0) 
            
     d_W_fc1 = weight_variable([100, 2])
     d_b_fc1 = bias_variable([2])
     d_logits = tf.matmul(d_h_fc0, d_W_fc1) + d_b_fc1
            
     self.domain_pred = tf.nn.softmax(d_logits)
     self.domain_loss = tf.nn.softmax_cross_entropy_with_logits(d_logits, self.domain)
```

I appreciate your help!

**Comment by ChengtaoLi** [comment]
Hi, we directly take the mean of the feature and then the rest of the structure of DAN is the same as the original single-sample adversary, except that now it only operates on data of batch_size=1 (which is the mean of the whole batch).

**Comment by alex-lew** [comment]
Thanks! I'm not sure I've implemented it the same way as you, but in my experiments, the distributional adversary (I'm using `s`, not `2s`) attains 100% accuracy and very small loss quite quickly. (Each batch, it sees the mean of the MNIST features and the mean of the MNIST-M features, and its loss is the mean of the two cross-entropy losses.) The loss of the per-sample adversary is typically around 0.1 or higher, whereas the loss of the distributional adversary is usually 1e-7 or similar. So, in my experiments, I don't see the distributional adversary having an effect -- I assume I am doing something wrong.

I don't want to take up your time with troubleshooting my code, but please do let me know when you provide your code for this experiment! Then I can check what I'm doing differently and correct my code :-) Thanks again!

**Comment by ChengtaoLi** [comment]
Hi, sorry for the late update -- we got distracted by some other things. The code for domain adaptation from MNIST to MNIST-M has been uploaded. Implementation based on https://github.com/pumpikano/tf-dann.

Since the distributional adversary is a stronger adversary, its loss at early phase of training becomes small. However, it is able to ``recover'', i.e., the generator is able to figure out ways to trick the distributional adversary after certain number of iterations.

Thanks for your interest and please let us know should you have other questions :)
